yt_dlp/extractor/vk.py

```diff
@@ -701,3 +702,131 @@ def _real_extract(self, url):
         return self.playlist_result(
             entries, post_id, join_nonempty(uploader, f'Wall post {post_id}', delim=' - '),
             clean_html(get_element_by_class('wall_post_text', webpage)))
+
+
+class VKPlayBaseIE(InfoExtractor):
+    def _extract_initial_state(self, url, video_id):
+        webpage = self._download_webpage(url, video_id)
+        initial_state_json = self._search_regex(r'id="initial-state"[^>]*>([^<]+)</script>', webpage, 'initial_state')
+        initial_state = self._parse_json(initial_state_json, video_id, fatal=False)
+        return initial_state
+
+    def _parse_playurls(self, playurls, video_id):
+        formats = []
+        for playurl in playurls:
+            if not playurl.get('url', None):
+                continue
+            elif '.m3u8' in playurl['url']:
+                try:
+                    formats.extend(self._extract_m3u8_formats(playurl['url'], video_id))
+                except Exception:
+                    pass
+            else:
+                formats.append(playurl)
+        return formats
```

Comment:
IMO a better way to do this is to pass the entire `stream_info`/`record_info` vars to the method.
It will reduce a bit more code duplication, and then we can use filtering in traversal to be safe (without try/except.)
Also, I noticed that there was a dash manifest not being extracted, added code to handle that:
```suggestion
    def _extract_formats(self, stream_info, video_id):
        formats = []
        for stream in traverse_obj(stream_info, (
                'data', 0, 'playerUrls', lambda _, v: url_or_none(v['url']) and v['type'])):
            url = stream['url']
            format_id = str_or_none(stream['type'])
            if format_id in ('hls', 'live_hls', 'live_playback_hls') or '.m3u8' in url:
                formats.extend(self._extract_m3u8_formats(url, video_id, m3u8_id=format_id, fatal=False))
            elif format_id in ('dash', 'live_dash', 'live_playback_dash'):
                formats.extend(self._extract_mpd_formats(url, video_id, mpd_id=format_id, fatal=False))
            else:
                formats.append({
                    'url': url,
                    'ext': 'mp4',
                    'format_id': format_id,
                    **parse_resolution(self._RESOLUTIONS.get(format_id)),
                })

        return formats
```
re: `**parse_resolution(self._RESOLUTIONS.get(format_id))`

in the current code, the direct http formats are being sorted behind m3u8, and they're not being sorted in any meaningful order amongst themselves. We could fix that by adding this class var:

```py
    _RESOLUTIONS = {
        'tiny': '256x144',
        'lowest': '426x240',
        'low': '640x360',
        'medium': '852x480',
        'high': '1280x720',
        'full_hd': '1920x1080',
        'quad_hd': '2560x1440',
    }
```

Reply:
The DASH url returns a json response instead of MPD xml and `_extract_mpd_formats()` doesn't work. I'll take a look how this can be handled.

Reply:
Hmm, just had a look at the DASH JSON response. IMO let's not even bother with it: it's only for livestreams and we don't have native support for live DASH anyways. It was my mistake to suggest that we extract it in the first place.

I think we should do this:

```py
            if format_id in ('hls', 'live_hls', 'live_playback_hls') or '.m3u8' in url:
                formats.extend(self._extract_m3u8_formats(url, video_id, m3u8_id=format_id, fatal=False))
            elif format_id in ('dash'):
                formats.extend(self._extract_mpd_formats(url, video_id, mpd_id=format_id, fatal=False))
            elif format_id in ('live_dash', 'live_playback_dash'):
                continue
            else:
```

Also, the `live_playback_hls` URLs seem to 404. We could move that to `elif ... continue` too, if you don't think it is useful. But keeping it is fine too, since it's non-fatal

Reply:
I am not sure where `live_playback_hls` might work, but it will be matched by `or '.m3u8' in url`, so I think it is just easier to leave it there.